File: FrontEnd/rema-partners/src/services/rating.service.ts
import axios from 'axios'
import type { Rating, RatingFormData, RatingReplyData } from '@/models/rating'

const API_BASE_URL = 'http://localhost:8080'

export const ratingService = {
    /**
     * Obtiene todas las valoraciones de un vendedor.
     * @param {string} sellerId - ID del vendedor.
     * @returns {Promise} Promesa con la lista de valoraciones.
     */
    async getRatingsBySeller(sellerId: string) {
        return axios.get(`${API_BASE_URL}/ratings/seller/${sellerId}`)
    },

    /**
     * Crea una nueva valoración.
     * @param {RatingFormData} ratingData - Datos de la valoración.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async createRating(ratingData: RatingFormData) {
        return axios.post(`${API_BASE_URL}/ratings/create`, ratingData)
    },

    /**
     * Actualiza una valoración existente.
     * @param {string} ratingId - ID de la valoración.
     * @param {RatingFormData} ratingData - Datos actualizados de la valoración.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async updateRating(ratingId: string, ratingData: RatingFormData) {
        return axios.put(`${API_BASE_URL}/ratings/${ratingId}`, ratingData)
    },

    /**
     * Obtiene la valoración que un usuario hizo a un vendedor específico.
     * @param {string} userId - ID del usuario.
     * @param {string} sellerId - ID del vendedor.
     * @returns {Promise} Promesa con la valoración encontrada.
     */
    async getUserRatingForSeller(userId: string, sellerId: string) {
        return axios.get(`${API_BASE_URL}/ratings/user/${userId}/seller/${sellerId}`)
    },

    /**
     * Añade o actualiza la respuesta del vendedor a una valoración.
     * @param {RatingReplyData} replyData - Datos de la respuesta.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async addReplyToRating(replyData: RatingReplyData) {
        return axios.post(`${API_BASE_URL}/ratings/reply`, replyData)
    },

    /**
     * Elimina una valoración.
     * @param {string} ratingId - ID de la valoración.
     * @returns {Promise} Promesa con la respuesta del servidor.
     */
    async deleteRating(ratingId: string) {
        return axios.delete(`${API_BASE_URL}/ratings/${ratingId}`)
    }
}

File: BackEnd/src/main/java/com/arsansys/RemaPartners/services/RatingService.java
package com.arsansys.RemaPartners.services;

import com.arsansys.RemaPartners.models.entities.RatingEntity;

import java.util.List;

/**
 * Servicio para la gestión de valoraciones de vendedores.
 */
public interface RatingService {

    /**
     * Obtiene todas las valoraciones de un vendedor.
     * 
     * @param sellerId ID del vendedor.
     * @return Lista de valoraciones.
     */
    List<RatingEntity> getRatingsBySeller(String sellerId);

    /**
     * Obtiene una valoración específica por su ID.
     * 
     * @param ratingId ID de la valoración.
     * @return La valoración si existe.
     */
    RatingEntity getRatingById(String ratingId);

    /**
     * Obtiene la valoración que un usuario hizo a un vendedor específico.
     * 
     * @param userId   ID del usuario.
     * @param sellerId ID del vendedor.
     * @return La valoración si existe.
     */
    RatingEntity getUserRatingForSeller(String userId, String sellerId);

    /**
     * Crea una nueva valoración.
     * 
     * @param rating Entidad de valoración a crear.
     * @return La valoración creada.
     */
    RatingEntity createRating(RatingEntity rating);

    /**
     * Actualiza una valoración existente.
     * 
     * @param ratingId ID de la valoración a actualizar.
     * @param rating   Datos actualizados.
     * @return La valoración actualizada.
     */
    RatingEntity updateRating(String ratingId, RatingEntity rating);

    /**
     * Añade o actualiza una respuesta a una valoración.
     * 
     * @param ratingId ID de la valoración.
     * @param reply    Texto de la respuesta.
     * @return La valoración actualizada con la respuesta.
     */
    RatingEntity addReplyToRating(String ratingId, String reply);

    /**
     * Elimina una valoración.
     * 
     * @param ratingId ID de la valoración a eliminar.
     */
    void deleteRating(String ratingId);
}
